troi/core.py

```diff
@@ -1,72 +1,65 @@
-    name: Override the algorithms that generate a playlist name and use this name instead.
-    desc: Override the algorithms that generate a playlist description and use this description instead.
-    min-recordings: The minimum number of recordings that must be present in a playlist to consider it complete.
-                    If it doesn't have sufficient numbers of tracks, ignore the playlist and don't submit it.
-                    Default: Off, a playlist with at least one track will be considere complete.
+    patch: the patch to run
+    pipeline_args: the arguments to pass to the entire pipeline
+    patch_args: the arguments passed to the topmost patch. this dict supports the following keys currently:
+        debug: Print debug information or not
+        args: Other parameters that might need to be passed to the patch [optional]
+        print: This option causes the generated playlist to be printed to stdout.
+        save: The save option causes the generated playlist to be saved to disk.
+        token: Auth token to use when using the LB API. Required for submitting playlists to the server.
+               See https://listenbrainz.org/profile to get your user token.
+        upload: Whether or not to submit the finished playlist to the LB server. Token must be set for this to work.
+        created-for: If this option is specified, it must give a valid user name and the
+                     TOKEN argument must specify a user who is whitelisted as a playlist bot at
+                     listenbrainz.org .
+        name: Override the algorithms that generate a playlist name and use this name instead.
+        desc: Override the algorithms that generate a playlist description and use this description instead.
+        min-recordings: The minimum number of recordings that must be present in a playlist to consider it complete.
+                        If it doesn't have sufficient numbers of tracks, ignore the playlist and don't submit it.
+                        Default: Off, a playlist with at least one track will be considere complete.
     """
-
-    if args is None:
-        args = []
-
-    patchname = patch
-    patches = troi.utils.discover_patches()
-    if patch not in patches:
-        print("Cannot load patch '%s'. Use the list command to get a list of available patches." % patch,
-              file=sys.stderr)
-        return None
-
-    patch = patches[patch](debug)
-
-    context = patch.parse_args.make_context(patchname, list(args))
-    pipelineargs = context.forward(patch.parse_args)
-
-    patch_args = {
-        "echo": echo,
-        "save": save,
-        "token": token,
-        "created_for": created_for, 
-        "upload": upload, 
-        "name": name,
-        "desc": desc,
-        "min_recordings": min_recordings
-    }
-
-    pipeline = patch.create(pipelineargs, patch_args)
+    patch_args = {**default_patch_args, **supplied_patch_args}
+    pipeline = patch.create(pipeline_args, patch_args)
```

Comment:
A few more things I was wondering was if we should merge patch_args and pipeline_args together. Because a quick look at the current patches and I see that even pipeline_args are only passed to top patch and not the entire pipeline elements.

I looked at the current uses of patch_args and I only saw it being used in YIM stuff and ForLoopElement which can probably be refactored to work with a single inputs dict.

Reply:
Done.